### PR TITLE
Fix cuda tests and error reporting in test discovery

### PR DESCRIPTION
**Pull request by sklam**
* Fix error reporting in test discovery.  Any error will trigger an `sys.exit(-1)` with the exception printed in stderr.
* Remove upper limit to test count in `test_runtests.py`.
* Fix cuda exception raising tests.
* Add source location reporting in cuda exception raising.  (#3325)

### CODE REVIEW

File: numba/tests/test_runtests.py [hunk 1]
```diff
@@ -24,46 +24,28 @@ def check_listing_prefix(self, prefix):
...
-            # inject into the error message for the user
-            try:
-                cmd = ['python', '-m', 'numba.runtests'] + list(args)
-                subprocess.check_output(cmd, stderr=subprocess.STDOUT)
-            except Exception as e:
-                msg = ("Test discovery has failed, the reported cause of the "
-                       " failure is:\n\n:")
-                indented  = '\n'.join(['\t' + x for x in
-                                       e.output.decode('UTF-8').splitlines()])
-                raise RuntimeError(msg + indented)
```
Comment (stuartarchibald) [inline numba/tests/test_runtests.py]: Nice to see that gone :)

File: numba/cuda/compiler.py [hunk 2]
```diff
@@ -582,9 +583,18 @@ def load_symbol(name):
                 tid = [load_symbol("tid" + i) for i in 'zyx']
                 ctaid = [load_symbol("ctaid" + i) for i in 'zyx']
                 code = excval.value
-                exccls, exc_args = self.call_helper.get_exception(code)
+                exccls, exc_args, loc = self.call_helper.get_exception(code)
+                # Prefix the exception message with the source location
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/cuda/compiler.py]: Yes, it'll be useful.  I just won't want to add too much changes in this PR.

[the remark this answers is not in the document]
Comment (sklam) [reply numba/cuda/compiler.py]: Added issue https://github.com/numba/numba/issues/3443 to track this

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cuda/compiler.py]: Great, thanks.

**Pull request merged**